terracotta/scripts/optimize_rasters.py

```diff
@@ -194,94 +280,48 @@ def optimize_rasters(raster_files: Sequence[Sequence[Path]],
+                        f'({i}/{len(raster_files_flat)})'
                     )
-
-            with contextlib.ExitStack() as es, warnings.catch_warnings():
-                warnings.filterwarnings('ignore', message='invalid value encountered.*')
-
-                src = es.enter_context(rasterio.open(str(input_file)))
-
-                if reproject:
-                    vrt = es.enter_context(_get_vrt(src, rs_method=rs_method))
-                else:
-                    vrt = src
-
-                profile = vrt.profile.copy()
-                profile.update(COG_PROFILE)
-
-                if in_memory is None:
-                    in_memory = vrt.width * vrt.height < IN_MEMORY_THRESHOLD
-
-                if in_memory:
-                    memfile = es.enter_context(MemoryFile())
-                    dst = es.enter_context(memfile.open(**profile))
-                else:
-                    tempraster = es.enter_context(TemporaryRasterFile(basedir=output_folder))
-                    dst = es.enter_context(rasterio.open(tempraster, 'w', **profile))
-
-                # iterate over blocks
-                windows = list(dst.block_windows(1))
-
-                for _, w in tqdm.tqdm(windows, desc='Reading', **sub_pbar_args):
-                    block_data = vrt.read(window=w, indexes=[1])
-                    dst.write(block_data, window=w)
-                    block_mask = vrt.dataset_mask(window=w).astype('uint8')
-                    dst.write_mask(block_mask, window=w)
-
-                # add overviews
-                if not in_memory:
-                    # work around bug mapbox/rasterio#1497
-                    dst.close()
-                    dst = es.enter_context(rasterio.open(tempraster, 'r+'))
-
-                max_overview_level = math.ceil(math.log2(max(
-                    dst.height // profile['blockysize'],
-                    dst.width // profile['blockxsize'],
-                    1
-                )))
-
-                if max_overview_level > 0:
-                    overviews = [2 ** j for j in range(1, max_overview_level + 1)]
-                    with tqdm.tqdm(desc='Creating overviews', total=1, **sub_pbar_args):
-                        dst.build_overviews(overviews, rs_method)
-
-                    dst.update_tags(ns='rio_overview', resampling=rs_method.value)
-
-                # copy to destination (this is necessary to push overviews to start of file)
-                with tqdm.tqdm(desc='Compressing', total=1, **sub_pbar_args):
-                    copy(
-                        dst, str(output_file), copy_src_overviews=True,
-                        compress=compression, **COG_PROFILE
-                    )
-
-            pbar.update(dst.height * dst.width)
+                    for i, input_file in enumerate(raster_files_flat, start=1)
+                ]
+
+                for future in concurrent.futures.as_completed(futures):
+                    future.result()  # Needed to throw any exceptions
```

Comment:
How about we do something like

```python
try:
    future.result()
except Exception as exc:
    raise RuntimeError(f'Error while optimizing file {filename}') from exc
```

Then we know which file acted up. For this you need a mapping of futures to filenames (you can use a dict, see [e.g. here](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example)).